### ISSUE TITLE
Cannot create new nodes, when internal state is frozen.

### ISSUE THREAD

**Issue by radmen** [phpmd/phpmd#328]
This is a bit tricky issue.

For start I know that this error is actually triggered by `pdepend/pdepend` package yet it happens when I'm using PHPMD so I decided to post this issue here.

I'm trying to run PHPMD on my PHP7 project. In most cases it fails with message:
```
Cannot create new nodes, when internal state is frozen.
```

I've tried to run PHPMD for some directories located inside my `src` dir and it worked. Yet again, when I tried to put them together separated by comma the fail message returned.

I'm not sure what's causing this problem. At first I've thought it's a problem of code written in PHP7 yet it seems to be unrelated (for single directories PHPMD works).

I'm using PHPMD in `2.3.2` version.
Tried to run binary in PHP `5.6.15`, and `7.0RC7`

**Comment by ravage84** [comment]
Hi @radmen , thanks for the issue, but it's a bit too vague.
Can you try to produce a minimum reproducable code/directory compilation?
Or could you fire up your debugger to help us investigate?

**Comment by radmen** [comment]
@ravage84 Unfortunatelly I cannot share the source code on which phpmd was failing. I'll try to create repo with failing PHPMD build.

**Comment by ravage84** [comment]
I totally understand that. Try what you can, we appreciate it.

**Comment by radmen** [comment]
@ravage84 I've managed to extract the failing parts from my project.

Here you can find repo: https://github.com/radmen/phpmd-nodes-freeze-error
Travis builds: https://travis-ci.org/radmen/phpmd-nodes-freeze-error

This code can be run under docker:

```
docker run -it -v "$PWD":/phpmd php:7.0 /phpmd/vendor/bin/phpmd /phpmd/src text /phpmd/phpmd.xml
```

**Comment by mikebronner** [comment]
I'm getting the same error. Is there a temporary work-around that I could use?

**Comment by quantumpacket** [comment]
I too am getting this error with a PHP 7.0RC7 project using PHPMD master. I'll narrow the code down and post something reproducible later today.

**Comment by rbairwell** [comment]
If it helps, I'm also getting the identical error (PHP7.0.0-dev Oct 22: the Ubuntu based "rasmus" dev Vagrant VM: running on a Windows 7 host with the project source files mounted from Windows) - but only on one of the two projects (one, a library project, with 3848 LOC, but it works okay on a Slim3 project which only has 399 LOC).

**Comment by quantumpacket** [comment]
I narrowed it down *(at least in my case)* to this example code. If I remove one method or the other the error does not appear. If I remove the `\stdClass` that too prevents the error.

**Errors**
```php
<?php

class Foo
{
    function bar() : \stdClass
    {

    }

    function baz()
    {

    }
}
```

**Success**
```php
<?php

class Foo
{
    function bar() : \stdClass
    {

    }
}
```

**Success**
```php
<?php

class Foo
{
    function baz()
    {

    }
}
```

**Success**
```php
<?php

class Foo
{
    function bar()
    {

    }

    function baz()
    {

    }
}
```

**Comment by quantumpacket** [comment]
@ravage84 Have you been able to reproduce this? Also, is there a way to disable this error for the time being?

**Comment by girvo** [comment]
Yeah that's what causing it in my case too, @quantumpacket

**Comment by quantumpacket** [comment]
@ravage84 is there a time-frame on when this may be fixed? It essentially breaks PHPMD in PHP7 applications using return type declarations so no other PHPMD warnings are triggered, and also as a side effect is piling up errors for every single modification to a file in PhpStorm. I'm easily getting 300+ errors over the course of editing a file and the only way to fix it is to comment out any return type declarations.

**Comment by shssoichiro** [comment]
Bump. Experiencing this issue as well with PHP7. I've had to completely disable PHPMD for my current project.

**Comment by marco-msg-ferrari** [comment]
+1

**Comment by marco-msg-ferrari** [comment]
I've manged to solve the issue but it is a problem in pdepend.

https://github.com/pdepend/pdepend/compare/master...marco-msg-ferrari:master

I need help creating the regression test to submit the PR.

**Comment by kuczek** [comment]
+1

**Comment by phoenixrises** [comment]
An update to this (as I've hit the same issue and wasn't sure how far it had gotten) seems to be that @marco-msg-ferrari submitted a [pull request](https://github.com/pdepend/pdepend/pull/234/) to pdepend which appears to have been merged into master.

**Comment by carlowens** [comment]
Just to also confirm I get the same error using PHP 7.0.3-4+deb.sury.org~trusty+1.  Disabling phpmd for now.

**Comment by djlambert** [comment]
+1

**Comment by elhachmi** [comment]
Same error on php 7.0.3

**Comment by kleijnweb** [comment]
Even though that PR is merged, still getting this error. Installed with:

```
composer global require pdepend/pdepend='dev-master as 2.3'  phpmd/phpmd=dev-master
Changed current directory to /var/lib/jenkins/.composer
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing pdepend/pdepend (2.2.3)
  - Installing pdepend/pdepend (dev-master 99bd1af)
    Cloning 99bd1af095b9147c257512b61b2ddcd358d1bc4a

  - Removing phpmd/phpmd (2.3.2)
  - Installing phpmd/phpmd (dev-master cd9c0b7)
    Cloning cd9c0b7cebf7681c87f0b1493fea676114813b80
```

**Comment by elhachmi** [comment]
I resolve this error temporarily by this, that may not work for you, but for me is enough while waiting a real solution, you must not use php7 namespace grouped declaration.
```php
$filename = 'vendor/pdepend/pdepend/src/main/php/PDepend/Source/Language/PHP/PHPBuilder.php';
$str = file_get_contents($filename);
//I comment this line :)
$newStr = str_replace('$this->frozen = true;', '//$this->frozen = true;', $str);
file_put_contents($filename, $newStr);
```
I run this once in my local env and at each build
perhaps this bug is caused by function return type declaration, when using a custom type.

**Comment by dereuromark** [comment]
We also experienced the same thing in CircleCI, locally all seems fine.

**Comment by zazoomauro** [comment]
+1
